### PR TITLE
feat: update complexity rule for optional chaining & default values

### PR DESCRIPTION
**Pull request by lo1tuma**
Fixes: #18060

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[x] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

In addition to the existing patterns that increase complexity I’ve added the following:
* `ChainExpression` to increase the complexity by one for every optional chain expression that is encountered
* `AssignmentPattern` to increase the complexity by one for every default value assignment in function parameters or desturcturing expressions

### CODE REVIEW

**Comment by amareshsm** [comment]
In the documentation, can we add examples demonstrating the usage of ChainExpression and AssignmentPattern?
https://eslint.org/docs/latest/rules/complexity

File: lib/rules/complexity.js
Comment (mdjermanovic) [inline lib/rules/complexity.js:112]: This doesn't report the first example from https://github.com/eslint/eslint/issues/18060:

```js
/*eslint complexity: ["error", 2]*/

// still valid (calculated complexity = 2)
function a(x) {
    return x?.a?.b?.c?.d?.e?.f;
}

```

Because this is one `ChainExpression`. Perhaps we should count each `?.`?

**Comment by lo1tuma** [comment]
> In the documentation, can we add examples demonstrating the usage of ChainExpression and AssignmentPattern? https://eslint.org/docs/latest/rules/complexity

I’ve thought about this as well and figured out that the current documentation doesn’t cover several cases that the rule already handles as for example nullish coalescing. If have added two examples now, but I’m not sure if we should overhaul the whole documentation about the rule to explain exactly what increases branching.

**Comment by Tanujkanti4441** [comment]
Can you also decrease the length of PR's title to pass all the CI checks

File: lib/rules/complexity.js
Comment (lo1tuma) [reply lib/rules/complexity.js]: You are right. I’ve misunderstood how `ChainExpression` works. I’ve added a test-case and updated the implementation. Thank you!

File: tests/lib/rules/complexity.js
Comment (fasttime) [inline tests/lib/rules/complexity.js:132]: What about optional call expressions like `function a(b) { b?.(); }`? Those should also increase the complexity I think.

Comment (lo1tuma) [reply tests/lib/rules/complexity.js:132]: Good catch! I’ve added test cases and implemented handling of optional call expression.

**Review by fasttime — APPROVED** [review]
LGTM, thanks!

**Review by nzakas — APPROVED** [review]
LGTM. Looks like @mdjermanovic was specifically requested to review this, so waiting for that review before merging.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks!

**Pull request merged**
